File: camera/frame_counter.py
import time

class WeightedFramerateCounter:
    smoothing = 0.95
    startTime = 0
    framerate = 0

    def start(self):
        self.startTime = time.time()
        self.framerate = 0
        
    def tick(self):
        timeNow = time.time()
        if self.startTime == 0:
            self.startTime = timeNow
            return
        elapsed = 1.0 / (timeNow - self.startTime)
        self.startTime = timeNow
        self.framerate = (self.framerate * self.smoothing) + (elapsed * (1.0 - self.smoothing))

    def getFramerate(self):
        return self.framerate 

File: camera/c_plus_plus/frame_counter.cpp
#include "frame_counter.h"

namespace Capstone {

    void WeightedFramerateCounter::startTime() {
        start = std::chrono::high_resolution_clock::now().time_since_epoch();
        frameRate = 0.0;
    }

    void WeightedFramerateCounter::tick() {
        std::chrono::duration<double> timeNow = std::chrono::high_resolution_clock::now().time_since_epoch();
        if (start == std::chrono::duration<double>::zero()) {
            start = timeNow;
            return;
        }
        double elapsed = 1.0 / (timeNow - start).count();
        start = timeNow;
        frameRate = (frameRate * smoothing) + (elapsed * (1.0 * smoothing));
    }

    float WeightedFramerateCounter::getFrameRate() {
        return frameRate;
    }
}
